zkevm-circuits/src/evm_circuit/execution/add.rs

```diff
@@ -0,0 +1,223 @@
+use crate::{
+    evm_circuit::{
+        execution::{
+            bus_mapping_tmp::{Block, Call, ExecStep, Transaction},
+            ExecutionGadget,
+        },
+        step::ExecutionResult,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StateTransition, Transition::Delta,
+            },
+            math_gadget::{PairSelectGadget, WordAdditionGadget},
+            select,
+        },
+    },
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+// AddGadget verifies ADD and SUB at the same time by an extra swap flag,
+// when it's ADD, we annotate stack as [a, b, ...] and [c, ...],
+// when it's SUB, we annotate stack as [c, b, ...] and [a, ...].
+// Then we verify if a + b is equal to c.
+#[derive(Clone, Debug)]
+pub(crate) struct AddGadget<F> {
+    same_context: SameContextGadget<F>,
+    word_addition: WordAdditionGadget<F>,
+    is_sub: PairSelectGadget<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for AddGadget<F> {
+    const NAME: &'static str = "ADD";
+
+    const EXECUTION_RESULT: ExecutionResult = ExecutionResult::ADD;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let word_addition = WordAdditionGadget::construct(cb);
+
+        // Swap b and c if it's SUB
+        let is_sub = PairSelectGadget::construct(
+            cb,
+            opcode.expr(),
+            OpcodeId::SUB.expr(),
+            OpcodeId::ADD.expr(),
+        );
+
+        // ADD: Pop a and b from the stack, push c on the stack
+        // SUB: Pop c and b from the stack, push a on the stack
+        let WordAdditionGadget { a, b, c, .. } = &word_addition;
+        cb.stack_pop(select::expr(is_sub.expr().0, c.expr(), a.expr()));
+        cb.stack_pop(b.expr());
+        cb.stack_push(select::expr(is_sub.expr().0, a.expr(), c.expr()));
```

Comment:
The `is_sub` swaps `a` and `c` instead of `b` and `c` ([reference](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/op_execution/arithmetic/add.rs#L108-L110) to previous impl)